RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,509 @@
+unspecified elements).
+
+To support Calculus functions on sparse tensors, we'll propose adding a
+fill-value property to PyTorch sparse tensors that will represent the
+values of unspecified sparse tensor elements.  While doing so, we also
+need to consider
+- how the non-zero fill-value affects the result of linear algebra
+  operations such as matrix multiplication, etc.
+- how to define a fill-value for hybrid tensors (tensors with sparse
+  and dense dimensions),
+- how PyTorch autograd support needs to deal with a defined
+  fill-value.
+
+
+## Proposal
+
+This proposal is about enabling sparse tensor support for Calculus
+domain while preserving the existing functionality for Linear Algebra
+domain, and allowing extensions to Graph domain.
+
+In the following, we describe the sparse tensor fill-value feature
+with examples to sparse tensors in COO storage format.  When a new
+sparse tensor storage format is introduced to PyTorch, the same
+semantics apply to the new format.
+
+1.  We propose to extend sparse tensor constructors with a keyword
+    argument `fill_value` that users can use to define the value for
+    unspecified elements of the constructed sparse tensor.
+
+    For instance, the Python signature of `sparse_coo_tensor` would be
+
+    ```python
+    torch.sparse_coo_tensor(indices, values, size=None, fill_value=None, dtype=None, device=None, requires_grad=False)
+    ```
+
+2.  The default fill-value is `0`.
+
+    This choice is consistent with the interpretation of unspecified
+    element in Linear Algebra domain.
+
+3.  PyTorch functions that have `layout` argument, may use different
+    fill-value when constructing a sparse tensor as defined by the
+    following table:
+
+    | Function              | Fill-value of returned sparse tensor |
+    | :-------------------- | :----------------------------------- |
+    | `torch.arange`        | 0                                    |
+    | `torch.empty`         | 0                                    |
+    | `torch.empty_like`    | 0                                    |
+    | `torch.empty_strided` | N/A                                  |
+    | `torch.eye`           | 0                                    |
+    | `torch.full`          | same as `fill_value` argument        |
+    | `torch.full_like`     | same as `fill_value` argument        |
+    | `torch.linspace`      | 0                                    |
+    | `torch.logspace`      | 0                                    |
+    | `torch.ones`          | 1                                    |
+    | `torch.ones_like`     | 1                                    |
+    | `torch.range`         | 0                                    |
+    | `torch.zeros`         | 0                                    |
+    | `torch.zeros_like`    | 0                                    |
+
+    For discussion: should we introduce the `fill_value` argument to
+    functions `empty` and `empty_like`?
```

Comment:
I'd prefer not doing that, given that it'd be a keyword that does not apply to dense tensors. One can easily set the fill value after construction.

Also, `empty` gives undefined values by definition, it just reserves a block of memory, so having already-defined fill values seems not needed.

Reply:
I now agree on not introducing `fill_value` to `empty` because the tensor content is arbitrary (in the sense that the content is not initialized), hence the `fill_value` content should be also arbitrary.

Re "Also, empty gives undefined values by definition...": I think by "undefined value" you mean "arbitrary value" as defined above. In this proposal, we are using the term "undefined fill-value" that has a different meaning: undefined fill-value is a not-a-value, like `None` in Python is sometimes used to represent the situation when a value has not been specified.

That said, I think the proposal should use "unspecified fill-value" (any better idea?) instead of "undefined fill-value" which is too similar to "undefined value" that has a different meaning: https://en.wikipedia.org/wiki/Undefined_value

Reply:
> I think by "undefined value" you mean "arbitrary value" as defined above.

Yes indeed. 

> That said, I think the proposal should use "unspecified fill-value" (any better idea?) instead of "undefined fill-value" 

Agreed, sounds like the clearest terminology to me.

Reply:
Second thought: " the unspecified fill-value can be specified..." sounds strange. So, a new suggestion is to use "indefinite fill-value" or "indeterminate fill-value".

Reply:
Using "indefinite fill-value".